File: backend/app/models/provider.py
from __future__ import annotations

from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, Integer, String, text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, relationship
from uuid import UUID

from app.db.types import JSONBCompat

from .base import Base, TimestampMixin, UUIDPrimaryKeyMixin
from .provider_preset import ProviderPreset
from .provider_allowed_user import ProviderAllowedUser


class Provider(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Database model storing provider metadata and configuration."""

    __tablename__ = "providers"

    provider_id: Mapped[str] = Column(String(50), unique=True, nullable=False, index=True)
    name: Mapped[str] = Column(String(100), nullable=False)
    base_url: Mapped[str] = Column(String(255), nullable=False)
    transport: Mapped[str] = Column(String(16), nullable=False, server_default=text("'http'"))
    provider_type: Mapped[str] = Column(
        String(16), nullable=False, server_default=text("'native'"), default="native"
    )
    sdk_vendor: Mapped[str | None] = Column(
        String(32),
        nullable=True,
        doc="When transport='sdk', identifies which official SDK implementation to use (e.g. openai/google/claude).",
    )
    weight: Mapped[float] = Column(Float, nullable=False, server_default=text("1.0"))
    region: Mapped[str | None] = Column(String(50), nullable=True)
    cost_input: Mapped[float | None] = Column(Float, nullable=True)
    cost_output: Mapped[float | None] = Column(Float, nullable=True)
    billing_factor: Mapped[float] = Column(
        Float,
        nullable=False,
        server_default=text("1.0"),
    )
    max_qps: Mapped[int | None] = Column(Integer, nullable=True)
    retryable_status_codes = Column(JSONBCompat(), nullable=True)
    custom_headers = Column(JSONBCompat(), nullable=True)
    models_path: Mapped[str] = Column(String(100), nullable=False, server_default=text("'/v1/models'"))
    messages_path: Mapped[str | None] = Column(String(100), nullable=True)
    chat_completions_path: Mapped[str] = Column(
        String(100), nullable=False, server_default=text("'/v1/chat/completions'"), default="/v1/chat/completions"
    )
    responses_path: Mapped[str | None] = Column(String(100), nullable=True)
    static_models = Column(JSONBCompat(), nullable=True)
    supported_api_styles = Column(JSONBCompat(), nullable=True)
    status: Mapped[str] = Column(String(16), nullable=False, server_default=text("'healthy'"))
    audit_status: Mapped[str] = Column(
        String(24),
        nullable=False,
        server_default=text("'pending'"),
        default="pending",
        doc="审核状态：pending/testing/approved/rejected/approved_limited",
    )
    operation_status: Mapped[str] = Column(
        String(16),
        nullable=False,
        server_default=text("'active'"),
        default="active",
        doc="运营状态：active/paused/offline",
    )
    probe_enabled: Mapped[bool] = Column(
        Boolean,
        nullable=False,
        server_default=text("true"),
        default=True,
        doc="是否开启自动探针/巡检",
    )
    probe_interval_seconds: Mapped[int | None] = Column(
        Integer,
        nullable=True,
        doc="自定义探针/巡检间隔（秒），为空则使用全局默认",
    )
    probe_model: Mapped[str | None] = Column(
        String(100),
        nullable=True,
        doc="探针使用的模型 ID（可选）",
    )
    last_check = Column(DateTime(timezone=True), nullable=True)
    metadata_json = Column("metadata", JSONBCompat(), nullable=True)

    # Ownership and visibility metadata.
    # When owner_id is NULL and visibility is "public", the provider is part of
    # the global shared pool. When owner_id is set and visibility is "private",
    # the provider is a user-scoped private provider.
    owner_id: Mapped[UUID | None] = Column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    preset_uuid: Mapped[UUID | None] = Column(
        PG_UUID(as_uuid=True),
        ForeignKey("provider_presets.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    visibility: Mapped[str] = Column(
        String(16),
        nullable=False,
        server_default=text("'public'"),
        default="public",
    )

    api_keys: Mapped[list["ProviderAPIKey"]] = relationship(
        "ProviderAPIKey",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    models: Mapped[list["ProviderModel"]] = relationship(
        "ProviderModel",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    api_key_restrictions: Mapped[list["APIKeyAllowedProvider"]] = relationship(
        "APIKeyAllowedProvider",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="selectin",
    )
    shared_users: Mapped[list["ProviderAllowedUser"]] = relationship(
        "ProviderAllowedUser",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
        lazy="selectin",
    )
    owner: Mapped["User"] = relationship("User")
    preset: Mapped[ProviderPreset | None] = relationship(
        "ProviderPreset", back_populates="providers", foreign_keys=[preset_uuid]
    )
    audit_logs: Mapped[list["ProviderAuditLog"]] = relationship(
        "ProviderAuditLog",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    test_records: Mapped[list["ProviderTestRecord"]] = relationship(
        "ProviderTestRecord",
        back_populates="provider",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )

    @property
    def preset_id(self) -> str | None:
        if self.preset is None:
            return None
        return self.preset.preset_id

    @property
    def shared_user_ids(self) -> list[UUID]:
        """辅助属性：返回已授权的用户 ID 列表。"""
        return [link.user_id for link in self.shared_users or []]


__all__ = ["Provider"]

File: frontend/http/provider.ts
import { httpClient } from './client';

// 提供商类型定义
export type ProviderVisibility = 'public' | 'private' | 'restricted';
export type ProviderType = 'native' | 'aggregator';
export type TransportType = 'http' | 'sdk';
export type SdkVendor = string;
export type ProviderStatus = 'healthy' | 'degraded' | 'down';
export type ProviderAuditStatus = 'pending' | 'testing' | 'approved' | 'approved_limited' | 'rejected';
export type ProviderOperationStatus = 'active' | 'paused' | 'offline';

// API Key 接口
export interface ProviderApiKey {
  key: string;
  weight: number;
  max_qps: number;
  label: string;
}

// 提供商接口
export interface Provider {
  id: string;
  provider_id: string;
  name: string;
  base_url: string;
  transport: TransportType;
  provider_type: ProviderType;
  sdk_vendor: SdkVendor | null;
  visibility: ProviderVisibility;
  owner_id: string | null;
  status: ProviderStatus;
  weight: number;
  region: string | null;
  cost_input: number;
  cost_output: number;
  billing_factor: number;
  max_qps: number | null;
  retryable_status_codes: number[] | null;
  custom_headers: Record<string, string> | null;
  models_path: string;
  messages_path: string | null;
  chat_completions_path: string;
  responses_path: string | null;
  supported_api_styles: string[] | null;
  static_models: any[] | null;
  shared_user_ids?: string[] | null;
  api_keys?: ProviderApiKey[];
  audit_status?: ProviderAuditStatus;
  operation_status?: ProviderOperationStatus;
  latest_test_result?: ProviderTestResult | null;
  probe_enabled?: boolean | null;
  probe_interval_seconds?: number | null;
  probe_model?: string | null;
  last_check?: number | string | null;
  created_at: string;
  updated_at: string;
}

export interface ProviderTestRequest {
  mode?: 'auto' | 'custom' | 'cron';
  remark?: string | null;
  input_text?: string | null;
}

export interface ProviderTestResult {
  id: string;
  provider_id: string;
  mode: string;
  success: boolean;
  summary?: string | null;
  probe_results?: any;
  latency_ms?: number | null;
  error_code?: string | null;
  cost?: number | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProviderAuditLog {
  id: string;
  provider_id: string;
  action: string;
  from_status?: string | null;
  to_status?: string | null;
  operation_from_status?: string | null;
  operation_to_status?: string | null;
  operator_id?: string | null;
  remark?: string | null;
  test_record_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface UpdateProbeConfigRequest {
  probe_enabled?: boolean;
  probe_interval_seconds?: number | null;
  probe_model?: string | null;
}

export interface ProviderModelValidationResult {
  model_id: string;
  success: boolean;
  latency_ms?: number | null;
  error_message?: string | null;
  timestamp?: string | null;
}

// Provider 模型接口（来自 `/providers/{id}/models`）
// 注意：后端已经对不同厂商的模型做了一层标准化：
// - 顶层字段是统一的（model_id / provider_id / family / display_name / context_length / capabilities / pricing / metadata / meta_hash）
// - metadata 中保留了各个厂商原始的模型信息（如 OpenAI 的 id / created / owned_by 等）
export interface ModelMetadata {
  id?: string;
  object?: string;
  created?: number;
  owned_by?: string;
  // 其它厂商自定义字段
  [key: string]: any;
}

export interface Model {
  model_id: string;
  provider_id: string;
  family: string;
  display_name: string;
  context_length: number;
  // 后端返回的是 Enum 的字符串值，例如 "chat" / "embedding" 等
  capabilities: string[];
  pricing: Record<string, number> | null;
   // 可选的模型别名，用于将长版本 ID 映射为更易记的短名称
  alias?: string | null;
  metadata?: ModelMetadata | null;
  meta_hash?: string | null;
}

// 管理端：provider+model 维度的计费配置
export interface ProviderModelPricing {
  provider_id: string;
  model_id: string;
  pricing: Record<string, number> | null;
}

// provider+model 维度的别名映射配置
export interface ProviderModelAlias {
  provider_id: string;
  model_id: string;
  alias: string | null;
}

export interface ProviderSharedUsersResponse {
  provider_id: string;
  visibility: ProviderVisibility;
  shared_user_ids: string[];
}

export interface UpdateProviderSharedUsersRequest {
  user_ids: string[];
}

export interface SDKVendorsResponse {
  vendors: SdkVendor[];
  total: number;
}

// 模型列表响应
export interface ModelsResponse {
  models: Model[];
  total: number;
}

// 健康状态接口
export interface HealthStatus {
  status: ProviderStatus;
  last_check: number;
  metadata: Record<string, any>;
}

// 提供商指标接口
export interface ProviderMetrics {
  logical_model: string;
  provider_id: string;
  latency_p95_ms: number;
  latency_p99_ms: number;
  error_rate: number;
  success_qps_1m: number;
  total_requests_1m: number;
  last_updated: number;
  status: ProviderStatus;
  // 额外字段用于计算
  avg_latency_ms: number;
  success_rate: number;
  total_requests: number;
  total_failures: number;
  window_start: number;
}

// 指标响应
export interface MetricsResponse {
  metrics: ProviderMetrics[];
}

// 用户可用提供商响应
export interface UserAvailableProvidersResponse {
  private_providers: Provider[];
  shared_providers: Provider[];
  public_providers: Provider[];
  total: number;
}

// 创建私有提供商请求
export interface CreatePrivateProviderRequest {
  name: string;
  base_url: string;
  api_key: string;
  provider_type?: ProviderType;
  transport?: TransportType;
  sdk_vendor?: SdkVendor;
  weight?: number;
  region?: string;
  cost_input?: number;
  cost_output?: number;
  max_qps?: number;
  retryable_status_codes?: number[];
  custom_headers?: Record<string, string>;
  models_path?: string;
  messages_path?: string;
  chat_completions_path?: string;
  responses_path?: string;
  supported_api_styles?: string[];
  static_models?: any[];
}

// 更新私有提供商请求
export interface UpdatePrivateProviderRequest {
  name?: string;
  base_url?: string;
  provider_type?: ProviderType;
  transport?: TransportType;
  sdk_vendor?: SdkVendor;
  weight?: number;
  region?: string;
  cost_input?: number;
  cost_output?: number;
  max_qps?: number;
  retryable_status_codes?: number[];
  custom_headers?: Record<string, string>;
  models_path?: string;
  messages_path?: string;
  chat_completions_path?: string;
  responses_path?: string;
  supported_api_styles?: string[];
  static_models?: any[];
}

// 提供商服务
export const providerService = {
  /**
   * 获取用户可用的提供商列表（私有 + 公共）
   */
  getUserAvailableProviders: async (
    userId: string,
    visibility?: 'all' | 'private' | 'public' | 'shared'
  ): Promise<UserAvailableProvidersResponse> => {
    const params = visibility ? { visibility } : {};
    const response = await httpClient.get(`/users/${userId}/providers`, { params });
    return response.data;
  },

  /**
   * 获取用户的私有提供商列表
   */
  getUserPrivateProviders: async (userId: string): Promise<Provider[]> => {
    const response = await httpClient.get(`/users/${userId}/private-providers`);
    return response.data;
  },

  /**
   * 创建私有提供商
   */
  createPrivateProvider: async (
    userId: string,
    data: CreatePrivateProviderRequest
  ): Promise<Provider> => {
    const response = await httpClient.post(
      `/users/${userId}/private-providers`,
      data
    );
    return response.data;
  },

  /**
   * 更新私有提供商
   */
  updatePrivateProvider: async (
    userId: string,
    providerId: string,
    data: UpdatePrivateProviderRequest
  ): Promise<Provider> => {
    const response = await httpClient.put(
      `/users/${userId}/private-providers/${providerId}`,
      data
    );
    return response.data;
  },

  /**
   * 删除私有提供商
   */
  deletePrivateProvider: async (
    userId: string,
    providerId: string
  ): Promise<void> => {
    await httpClient.delete(`/users/${userId}/private-providers/${providerId}`);
  },

  /**
   * 获取指定提供商信息
   */
  getProvider: async (providerId: string): Promise<Provider> => {
    const response = await httpClient.get(`/providers/${providerId}`);
    return response.data;
  },

  /**
   * 获取提供商模型列表
   */
  getProviderModels: async (providerId: string): Promise<ModelsResponse> => {
    const response = await httpClient.get(`/providers/${providerId}/models`);
    return response.data;
  },

  /**
   * 检查提供商健康状态
   */
  checkProviderHealth: async (providerId: string): Promise<HealthStatus> => {
    const response = await httpClient.get(`/providers/${providerId}/health`);
    return response.data;
  },

  /**
   * 获取提供商路由指标（实时快照）
   */
  getProviderMetrics: async (
    providerId: string,
    logicalModel?: string
  ): Promise<MetricsResponse> => {
    const params = logicalModel ? { logical_model: logicalModel } : {};
    const response = await httpClient.get(`/providers/${providerId}/metrics`, { params });
    return response.data;
  },

  /**
   * 获取指定 provider+model 的别名映射配置
   */
  getProviderModelAlias: async (
    providerId: string,
    modelId: string
  ): Promise<ProviderModelAlias> => {
    const response = await httpClient.get(
      `/providers/${providerId}/models/${encodeURIComponent(modelId)}/mapping`
    );
    return response.data;
  },

  /**
   * 更新指定 provider+model 的别名映射配置
   */
  updateProviderModelAlias: async (
    providerId: string,
    modelId: string,
    data: { alias?: string | null }
  ): Promise<ProviderModelAlias> => {
    const response = await httpClient.put(
      `/providers/${providerId}/models/${encodeURIComponent(modelId)}/mapping`,
      data
    );
    return response.data;
  },

  /**
   * 获取指定 provider+model 的计费配置（管理员接口）
   */
  getProviderModelPricing: async (
    providerId: string,
    modelId: string
  ): Promise<ProviderModelPricing> => {
    const response = await httpClient.get(
      `/admin/providers/${providerId}/models/${encodeURIComponent(modelId)}/pricing`
    );
    return response.data;
  },

  /**
   * 查询或更新私有 Provider 的授权用户列表
   */
  getProviderSharedUsers: async (
    userId: string,
    providerId: string
  ): Promise<ProviderSharedUsersResponse> => {
    const response = await httpClient.get(
      `/users/${userId}/private-providers/${providerId}/shared-users`
    );
    return response.data;
  },

  updateProviderSharedUsers: async (
    userId: string,
    providerId: string,
    data: UpdateProviderSharedUsersRequest
  ): Promise<ProviderSharedUsersResponse> => {
    const response = await httpClient.put(
      `/users/${userId}/private-providers/${providerId}/shared-users`,
      data
    );
    return response.data;
  },

  /**
   * 更新指定 provider+model 的计费配置（管理员接口）
   */
  updateProviderModelPricing: async (
    providerId: string,
    modelId: string,
    data: { input?: number; output?: number } | null
  ): Promise<ProviderModelPricing> => {
    const response = await httpClient.put(
      `/admin/providers/${providerId}/models/${encodeURIComponent(modelId)}/pricing`,
      data
    );
    return response.data;
  },

  /**
   * 获取已注册的 SDK 厂商列表
   */
  getSdkVendors: async (): Promise<SDKVendorsResponse> => {
    const response = await httpClient.get('/providers/sdk-vendors');
    return response.data;
  },

  /**
   * 管理员触发 Provider 测试
   */
  adminTestProvider: async (
    providerId: string,
    payload?: ProviderTestRequest
  ): Promise<ProviderTestResult> => {
    const response = await httpClient.post(
      `/admin/providers/${providerId}/test`,
      payload ?? {}
    );
    return response.data;
  },

  /**
   * 审核通过/限速通过 Provider
   */
  approveProvider: async (
    providerId: string,
    payload?: { remark?: string; limit_qps?: number | null; limited?: boolean }
  ): Promise<Provider> => {
    const path = payload?.limited ? "approve-limited" : "approve";
    const { limited: _omitted, ...body } = payload ?? {};
    const response = await httpClient.post(
      `/admin/providers/${providerId}/${path}`,
      body ?? {}
    );
    return response.data;
  },

  /**
   * 审核拒绝 Provider（remark 必填）
   */
  rejectProvider: async (
    providerId: string,
    payload: { remark: string }
  ): Promise<Provider> => {
    const response = await httpClient.post(
      `/admin/providers/${providerId}/reject`,
      payload
    );
    return response.data;
  },

  /**
   * 更新运营状态：暂停/恢复/下线
   */
  updateOperationStatus: async (
    providerId: string,
    action: 'pause' | 'resume' | 'offline',
    payload?: { remark?: string | null }
  ): Promise<Provider> => {
    const response = await httpClient.post(
      `/admin/providers/${providerId}/${action}`,
      payload ?? {}
    );
    return response.data;
  },

  /**
   * 管理员查看测试记录
   */
  getProviderTests: async (
    providerId: string,
    params?: { limit?: number }
  ): Promise<ProviderTestResult[]> => {
    const response = await httpClient.get(
      `/admin/providers/${providerId}/tests`,
      { params }
    );
    return response.data;
  },

  /**
   * 管理员查看审核日志
   */
  getProviderAuditLogs: async (
    providerId: string,
    params?: { limit?: number }
  ): Promise<ProviderAuditLog[]> => {
    const response = await httpClient.get(
      `/admin/providers/${providerId}/audit-logs`,
      { params }
    );
    return response.data;
  },

  /**
   * 更新探针配置（启停/频率/模型）
   */
  updateProbeConfig: async (
    providerId: string,
    payload: UpdateProbeConfigRequest
  ): Promise<Provider> => {
    const response = await httpClient.put(
      `/admin/providers/${providerId}/probe-config`,
      payload
    );
    return response.data;
  },

  /**
   * 验证静态模型可用性（轻量 chat 调用）
   */
  validateProviderModels: async (
    providerId: string,
    params?: { limit?: number }
  ): Promise<ProviderModelValidationResult[]> => {
    const response = await httpClient.post(
      `/admin/providers/${providerId}/validate-models`,
      null,
      { params }
    );
    return response.data;
  },
};
